consensus/src/round_manager.rs

```diff
@@ -292,61 +292,40 @@ impl RoundManager {
-        self.process_proposed_block(block).await
-    }
-
-    /// The function is responsible for processing the incoming proposals and the Quorum
-    /// Certificate.
-    /// 1. sync up to the SyncInfo including committing to the committed state the HCC carries
-    /// and fetch all the blocks from the committed state to the HQC
-    /// 2. forwarding the proposals to the ProposerElection queue,
-    /// which is going to eventually trigger one winning proposal per round
-    async fn pre_process_proposal(&mut self, proposal_msg: ProposalMsg) -> Result<Block> {
         trace_event!("round_manager::pre_process_proposal", {"block", proposal_msg.proposal().id()});
-        // RoundState is going to be updated with all the proposal certificates later,
-        // but it's known that the round_state's round is not going to decrease so we can already
-        // filter out the proposals from old rounds.
-        let current_round = self.round_state.current_round();
-        ensure!(
-            proposal_msg.round() >= current_round,
-            "[RoundManager] Proposal round {} is less than current round {}",
-            proposal_msg.round(),
-            current_round,
-        );
-        ensure!(
-            self.proposer_election
-                .is_valid_proposal(proposal_msg.proposal()),
-            "[RoundManager] Proposer {} for block {} is not a valid proposer for this round",
+        self.ensure_round(
+            proposal_msg.proposal().round(),
+            proposal_msg.sync_info(),
             proposal_msg.proposer(),
-            proposal_msg.proposal()
-        );
-        self.sync_up(proposal_msg.sync_info(), proposal_msg.proposer(), true)
-            .await
-            .context("[RoundManager] Process proposal")?;
-
-        // round_state may catch up with the SyncInfo, check again
-        let current_round = self.round_state.current_round();
-        ensure!(
-            proposal_msg.round() == current_round,
-            "[RoundManager] Proposal round doesn't match current round after sync"
-        );
-        Ok(proposal_msg.take_proposal())
+            true,
+        )
+        .await
+        .context("[RoundManager] Process proposal")?;
+        self.process_proposal(proposal_msg.take_proposal()).await
     }
 
-    /// The function makes sure that it brings the missing dependencies from the QC and LedgerInfo
-    /// of the given sync info and update the round_state with the certificates if succeed.
+    /// The function makes sure that it ensures the message_round equal to what we have locally,
+    /// brings the missing dependencies from the QC and LedgerInfo of the given sync info and
+    /// update the round_state with the certificates if succeed.
     /// Returns Error in case sync mgr failed to bring the missing dependencies.
     /// We'll try to help the remote if the SyncInfo lags behind and the flag is set.
-    pub async fn sync_up(
+    pub async fn ensure_round(
```

Comment:
ensure and sync? the function's name seems like this shouldn't expect a mutable self... or perhaps we have two functions?